src/main/scala/org/broadinstitute/dsde/firecloud/service/TSVFileSupport.scala

```diff
@@ -143,10 +147,38 @@ trait TSVFileSupport {
       val nameEntry = "attributeName" -> AttributeString(attributeName)
       def valEntry( attr: Attribute ) = "addUpdateAttribute" -> attr
       refTypeOpt match {
-        case Some(refType) => Map(upsertAttrOperation,nameEntry,valEntry(AttributeEntityReference(refType,value)))
+        case Some(refType) => Seq(Map(upsertAttrOperation,nameEntry,valEntry(AttributeEntityReference(refType,value))))
         case None => value match {
-          case "__DELETE__" => Map(removeAttrOperation,nameEntry)
-          case _ => Map(upsertAttrOperation,nameEntry,valEntry(AttributeString(value)))
+          case "__DELETE__" => Seq(Map(removeAttrOperation,nameEntry))
+          case value if modelSchema.isAttributeArray(value) => {
+            val listElements = value.parseJson.convertTo[JsArray].elements.toList
+            val listNameEntry = "attributeListName" -> AttributeString(attributeName)
+            def listValEntry( attr: Attribute ) = "newMember" -> attr
+
+            //if the list is empty, short-circuit and just replace any existing list with an empty list
+            if(listElements.isEmpty) {
+              Seq(Map(removeAttrOperation, nameEntry), Map(createAttrValueListOperation, nameEntry))
+            } else {
+              val addElements = listElements match {
+                case elements if elements.forall(_.isInstanceOf[JsString]) =>
+                  val elementsTyped: List[JsString] = elements.asInstanceOf[List[JsString]]
+                  elementsTyped.map(jsstr => Map(addListMemberOperation, listNameEntry, listValEntry(AttributeString(jsstr.value))))
+
+                case elements if elements.forall(_.isInstanceOf[JsNumber]) =>
+                  val elementsTyped: List[JsNumber] = elements.asInstanceOf[List[JsNumber]]
+                  elementsTyped.map(jsnum => Map(addListMemberOperation, listNameEntry, listValEntry(AttributeNumber(jsnum.value))))
+
+                case elements if elements.forall(_.isInstanceOf[JsBoolean]) =>
+                  val elementsTyped: List[JsBoolean] = elements.asInstanceOf[List[JsBoolean]]
+                  elementsTyped.map(jsbool => Map(addListMemberOperation, listNameEntry, listValEntry(AttributeBoolean(jsbool.value))))
+
+                case _ => throw new FireCloudException("Mixed-type entity attribute lists are not supported.")
+              }
+              val removeOldListOp = Seq(Map(removeAttrOperation, nameEntry))
+              removeOldListOp ++ addElements
+            }
+          }
```

Comment:
Could we refactor this block of code into its own function or maybe even multiple functions? Shorter methods whose names describe what's going on are nice I think.

Reply:
I broke this out along with some smaller pieces. 


I was also hoping to also use type parameters to slim this function down (so there aren't 3 extremely similar `case` branches), but I don't think it will work in the way that I wanted

This doesn't handle the mixed-type exception, but I was thinking something roughly like this (does not compile):

```
  def generateAttributeArrayOperations[T, U](attributeValue: String, attributeName: String): Seq[Map[String, Attribute]] = {
    val listElements = attributeValue.parseJson.convertTo[JsArray].elements.toList

    //if the list is empty, short-circuit and just replace any existing list with an empty list
    if(listElements.isEmpty) {
      Seq(Map(removeAttrOperation, nameEntry(attributeName)), Map(createAttrValueListOperation, nameEntry(attributeName)))
    } else {
      val elementsTyped: List[T] = listElements.asInstanceOf[List[T]]
      val elementsToAdd = elementsTyped.map(jsval => Map(addListMemberOperation, listNameEntry(attributeName), listValEntry(U(jsval.value))))
      
      val removeOldListOp = Seq(Map(removeAttrOperation, nameEntry(attributeName)))
      removeOldListOp ++ elementsToAdd
    }
  }
```